### DISCUSSION TITLE
How to use side_effect from pytest framework with pydantic.ValidationError class?

### DISCUSSION THREAD

**Discussion by leonartdoss** [pydantic/pydantic#6459] [Question] [answered]
Hi,

I'm trying to test my model validation exceptions with pytest.
When using the side_effect property (https://docs.python.org/3/library/unittest.mock.html#the-mock-class) I'm getting a TypeError, because the class pydantic.ValidationError does not have a constructor:

```python
validate_mock = mocker.patch(
        "my_package.MyClass.model_validate_json",
        side_effect=pydantic.ValidationError(...),
)

TypeError: constructor is not defined
```

How to face this issue? Would you have an idea?
Thank you

Best regards,
Leonardo

**Comment by mzamini92** [comment]
The ValidationError class in Pydantic does not have a constructor, so you can't use the side_effect property to pass it as an argument to the mock.patch() function. you can use the side_effect_functools.partial() function to create a partial function that takes the exception message as an argument. 
```
import pytest
import pydantic
from functools import partial


def test_model_validate_json():
    validate_mock = mocker.patch(
        "my_package.MyClass.model_validate_json",
        side_effect=partial(pydantic.ValidationError, "Invalid model"),
    )

    with pytest.raises(pydantic.ValidationError):
        MyClass.model_validate_json()
````

**Comment by leonartdoss** [comment]
I tried like this, but it doesn't work: 

```python
/usr/lib/python3.9/unittest/mock.py:1096: in _mock_call
    return self._execute_mock_call(*args, **kwargs)


def _execute_mock_call(self, /, *args, **kwargs):
        # separate from _increment_mock_call so that awaited functions are
        # executed separately from their call, also AsyncMock overrides this method
    
        effect = self.side_effect
        if effect is not None:
            if _is_exception(effect):
                raise effect
            elif not _callable(effect):
                result = next(effect)
                if _is_exception(result):
                    raise result
            else:
>               result = effect(*args, **kwargs)
E               TypeError: No constructor defined

/usr/lib/python3.9/unittest/mock.py:1157: TypeError
```

**Reply by mzamini92** [reply]
what if you create a custom validation error class called my_validation_error. The my_validation_error class takes a single positional argument, which is the error message. then use the wraps function to wrap the my_validation_error class and add a constructor that takes a single positional argument. Finally, use the mocker.patch() function to patch the model_validate_json method in the MyClass class. something like this:
```
import pytest
import pydantic
from functools import wraps


def my_validation_error(message):
    """A custom validation error class."""
    return pydantic.ValidationError(message)


@wraps(pydantic.ValidationError)
def wrapped_validation_error(message):
    """A wrapped validation error class."""
    return pydantic.ValidationError(message)


def test_model_validate_json():
    validate_mock = mocker.patch(
        "my_package.MyClass.model_validate_json",
        side_effect=wrapped_validation_error("Invalid model"),
    )

    with pytest.raises(pydantic.ValidationError):
        MyClass.model_validate_json()
```

**Comment by gleniosp** [comment]
Hi @gamaleonardo, maybe as a workaround you could use a function that throws the ValidationError as a `side_effect`, like:

```python
validate_mock = mocker.patch(
        "my_package.MyClass.model_validate_json",
        side_effect=lambda _: MyClass.model_validate_json(<some json that will fail the validation and cause a ValidationError>),
)
```

**Comment by nekeal** [comment, marked as answer]
You could use `ValidationError.from_exception_data()` method.

```python
from unittest.mock import Mock
from pydantic import ValidationError
func = Mock()
func.side_effect = ValidationError.from_exception_data("Invalid data", line_errors=[])
func()
```

**Reply by vcidst** [reply]
Looks like the `ValidationError.from_exception_data()` method was added in Pydantic 2.x, is there any way workaround for Pydantic 1.x?

**Reply by bmchilds** [reply]
For pydantic 1.x you can just do
func = Mock()
func.side_effect = ValidationError([], YOUR_MODEL)
func()

**Reply by evan-hitobito** [reply]
Thank you! This helps
